File: solutions_5658571765186560_1/Python/HannesF/D_Ominous_Omino.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

T = int(input())

for t in range(1, T+1):
    x, r, c = [int(i) for i in input().split()]
    if (x == 1  # trivial
            or (x == 2 and (r * c) % 2 == 0)
            ):
        print("Case #{}:".format(t), 'GABRIEL')
    elif ((r * c) % x != 0        # trivial
            or x >= 7           # chose second of samples given in problem discription
            or x > max(r, c)    # Xx1
            or x > r + c - 1    # T-shape does not fit
            or (x == r + c - 1 and r != 1 and c != 1) # T-shape can cut of min(r, c) - 1 < r + c - 1
            or (x == r + c - 1 and (r == 1 or c == 1)) # T-shape does not fit since x >= 3
            ):
        print("Case #{}:".format(t), 'RICHARD')
    # Special cases for 4 and 6 (short T-shape)
    # Note: does not happen for x == 5 (and cross or t-shape)
    elif ((x == 4 and min(r, c) == 2)    # remaining left and right are odd
            or (x == 6 and min(r, c) == 3) # remaining are 1 or 2 mod 3
            ):
        print("Case #{}:".format(t), 'RICHARD')

    else:       #
        # 3 <= x <= 6 and the peace can be placed that it touches only one side
        # (no small regions are cut off)
        print("Case #{}:".format(t), 'GABRIEL')  # possible

File: solutions_5658571765186560_0/C++/btzy/D-ominous-omino.cpp
#include<cstdio>
#include<algorithm>
using namespace std;
int tc,x,r,c;
int main(){
    scanf("%d\n",&tc);
    for(int ct=1;ct<=tc;++ct){
        scanf("%d%d%d",&x,&r,&c);
        if(r*c%x!=0||x>=7||x>max(r,c)||(x+1)/2>min(r,c)||(x==4&&min(r,c)==2)||(x==6&&min(r,c)==3))printf("Case #%d: RICHARD\n",ct);
        else printf("Case #%d: GABRIEL\n",ct);
    }
}
